// react-circular-slider/dist/components/Arc.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var d3_shape_1 = require("d3-shape");
var defaultStyle = {
    position: 'absolute',
    top: 0,
    left: 0,
};
exports.default = function (_a) {
    var _b = _a.radius, radius = _b === void 0 ? 80 : _b, _c = _a.startAngle, startAngle = _c === void 0 ? 0 : _c, _d = _a.endAngle, endAngle = _d === void 0 ? 360 : _d, _e = _a.padding, padding = _e === void 0 ? 10 : _e, _f = _a.thickness, thickness = _f === void 0 ? 20 : _f, style = _a.style, children = _a.children, mask = _a.mask, otherProps = __rest(_a, ["radius", "startAngle", "endAngle", "padding", "thickness", "style", "children", "mask"]);
    var arcData = d3_shape_1.arc()({
        innerRadius: radius - thickness,
        outerRadius: radius,
        startAngle: Math.PI * 2 * (startAngle / 360),
        endAngle: Math.PI * 2 * (endAngle / 360),
        padAngle: 0,
    });
    var size = (padding + radius) * 2;
    return (React.createElement("svg", __assign({ height: size, width: size, viewBox: "0 0 " + size + " " + size, style: __assign({}, defaultStyle, style) }, otherProps),
        mask && (React.createElement("defs", null,
            React.createElement("mask", { id: "arcmask" },
                React.createElement("rect", { width: size, height: size, fill: "white" }),
                mask))),
        React.createElement("g", { mask: "url(#arcmask)" },
            React.createElement("path", { d: arcData, transform: "translate(" + size / 2 + " " + size / 2 + ")" }),
            children)));
};
