// react-circular-slider/dist/components/Draggable.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var classnames = require("classnames");
var defaultStyle = {
    backgroundColor: 'black',
    borderRadius: '50%',
    boxSizing: 'border-box',
    cursor: 'pointer',
    height: 40,
    position: 'absolute',
    width: 40,
};
var Draggable = function (_a) {
    var children = _a.children, coordinates = _a.coordinates, _b = _a.pressed, pressed = _b === void 0 ? false : _b, _c = _a.rotation, rotation = _c === void 0 ? 0 : _c, _d = _a.size, size = _d === void 0 ? 40 : _d, style = _a.style;
    return (React.createElement("div", { style: __assign({}, defaultStyle, style, { width: size, height: size, top: coordinates ? coordinates.y : 0, left: coordinates ? coordinates.x : 0, transform: "translate(-50%, -50%) rotate(" + rotation + "deg)" }), className: classnames({ pressed: pressed }) }, children));
};
exports.default = Draggable;
